File: algorithm/go/firstround/111.go
package firstround

type TreeNode struct {
	Val         int
	Left, Right *TreeNode
}

func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	q := make([]*TreeNode, 0)
	q = append(q, root)
	res := 1
	for len(q) > 0 {
		length := len(q)
		for i := 0; i < length; i++ {
			node := q[0]
			q = q[1:]

			// 判断是否为叶子结点
			if node.Left == nil && node.Right == nil {
				return res
			}

			if node.Left != nil {
				q = append(q, node.Left)
			}

			if node.Right != nil {
				q = append(q, node.Right)
			}
		}
		res++
	}

	return -1
}

File: algorithm/cpp/firstround/111.cpp
#include <cstddef>
#include <algorithm>
#include <climits>
using namespace std;

struct TreeNode
{
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int x): val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
        int minDepth(TreeNode* root);
private:
        int postorder(TreeNode* root);
};

int Solution::minDepth(TreeNode* root)
{
    if(root==NULL) return 0;
    return postorder(root);
}

int Solution::postorder(TreeNode* root)
{
        if(root==NULL) return INT_MAX;

        if(root->left==NULL && root->right==NULL) return 1;
        return min(postorder(root->left), postorder(root->right))+1;
}

int main()
{
	return 0;
}

